File: coursehub/src/main/java/com/coursehub/service/CourseService.java
package com.coursehub.service;

import com.coursehub.dto.request.CreateCourseRequest;
import com.coursehub.dto.response.CourseResponse;
import com.coursehub.dto.response.PageResponse;
import com.coursehub.dto.response.LearningCourseResponse;
import org.springframework.web.multipart.MultipartFile;

import java.math.BigDecimal;
import java.util.UUID;

public interface CourseService {
    CourseResponse createCourse(UUID instructorId, CreateCourseRequest request);
    CourseResponse getCourseBySlug(String slug, UUID currentUserId);
    CourseResponse getCourseById(UUID courseId, UUID currentUserId);
    PageResponse<CourseResponse> getInstructorCourses(UUID instructorId, int page, int size);
    CourseResponse updateCourse(UUID instructorId, UUID courseId, CreateCourseRequest request);
    String uploadThumbnail(UUID instructorId, UUID courseId, MultipartFile file);
    void submitForReview(UUID instructorId, UUID courseId);
    void deleteCourse(UUID instructorId, UUID courseId);

    // Admin actions
    void approveCourse(UUID adminId, UUID courseId, String note);
    void rejectCourse(UUID adminId, UUID courseId, String note);
    void blockCourse(UUID adminId, UUID courseId, String note);

    // Public
    PageResponse<CourseResponse> searchCourses(String keyword, Long categoryId, String level,
                                               BigDecimal minPrice, BigDecimal maxPrice, BigDecimal rating,
                                               String language, String sortBy, int page, int size);

    CourseResponse getCourseDetailPublic(UUID courseId);
    LearningCourseResponse getLearningCourseContent(UUID userId, UUID courseId);
}

File: coursehub-frontend/src/services/courseService.ts
import axiosInstance from '../api/axiosInstance';
import type { ApiResponse } from '../types/auth.types';

export const courseService = {
  searchCourses: async (params: {
    keyword?: string;
    categoryId?: number | string;
    level?: string;
    minPrice?: number | string;
    maxPrice?: number | string;
    rating?: number | string;
    language?: string;
    sortBy?: string;
    page?: number;
    size?: number;
  }) => {
    const res = await axiosInstance.get<ApiResponse<any>>('/courses/search', { params });
    return res.data;
  },

  getCourseBySlug: async (slug: string) => {
    const res = await axiosInstance.get<ApiResponse<any>>(`/courses/public/${slug}`);
    return res.data;
  },

  createCourse: async (data: any) => {
    const res = await axiosInstance.post<ApiResponse<any>>('/instructor/courses', data);
    return res.data;
  },

  getInstructorCourses: async (page = 0, size = 12) => {
    const res = await axiosInstance.get<ApiResponse<any>>('/instructor/courses', {
      params: { page, size },
    });
    return res.data;
  },

  getCourseById: async (courseId: string) => {
    const res = await axiosInstance.get<ApiResponse<any>>(`/instructor/courses/${courseId}`);
    return res.data;
  },

  updateCourse: async (courseId: string, data: any) => {
    const res = await axiosInstance.put<ApiResponse<any>>(`/instructor/courses/${courseId}`, data);
    return res.data;
  },

  uploadThumbnail: async (courseId: string, file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    const res = await axiosInstance.post<ApiResponse<string>>(
      `/instructor/courses/${courseId}/thumbnail`,
      formData,
      { headers: { 'Content-Type': 'multipart-form-data' } }
    );
    return res.data;
  },

  submitForReview: async (courseId: string) => {
    const res = await axiosInstance.post<ApiResponse<void>>(`/instructor/courses/${courseId}/submit`);
    return res.data;
  },

  resubmitForReview: async (courseId: string) => {
    const res = await axiosInstance.put<ApiResponse<void>>(`/instructor/courses/${courseId}/resubmit`);
    return res.data;
  },

  deleteCourse: async (courseId: string) => {
    const res = await axiosInstance.delete<ApiResponse<void>>(`/instructor/courses/${courseId}`);
    return res.data;
  },

  // Admin moderation endpoints
  approveCourse: async (courseId: string, note?: string) => {
    const res = await axiosInstance.put<ApiResponse<void>>(`/admin/courses/${courseId}/approve`, null, {
      params: { note },
    });
    return res.data;
  },

  rejectCourse: async (courseId: string, note?: string) => {
    const res = await axiosInstance.put<ApiResponse<void>>(`/admin/courses/${courseId}/reject`, null, {
      params: { note },
    });
    return res.data;
  },
};
